src/main/scala/com/fulcrumgenomics/umi/ExtractUmisFromBam.scala

```diff
@@ -191,9 +193,44 @@ object ExtractUmisFromBam {
           case _ => None
         }
     }
+    // update any clipping information
+    updateClippingInformation(record=record, clippingAttribute=clippingAttribute, readStructureBases=readStructureBases)
     record.setReadString(basesAndQualities.map(_._1).mkString)
     record.setBaseQualityString(basesAndQualities.map(_._2).mkString)
     // return the concatenation of the molecular barcodes in sequencing order
     molecularBarcodeBases.mkString
   }
+
+
+  /**
+    * Update the clipping information produced by Picard's MarkIlluminaAdapters to account for any non-template bases
+    * that will be removed from the read.
+    */
+  private[umi] def updateClippingInformation(record: SAMRecord,
+                                             clippingAttribute: Option[String],
+                                             readStructureBases: Seq[(String, String, ReadSegment)]): Unit = {
+    clippingAttribute.map(tag => (tag, record.getIntegerAttribute(tag))) match {
+      case None =>
+      case Some((tag, null)) =>
+      case Some((tag, clippingPosition)) =>
+        var i = 0
+        var basesSoFar = 0
+        var newClippingPosition = 0
+        while (i < readStructureBases.length && basesSoFar < clippingPosition) {
```

Comment:
Maybe I'm missing something, but couldn't this whole loop be replaced with something like (assuming you pass in the ReadStructure as rs):
```scala
val minus = rs.takeWhile(_.offset < clippingPosition).filter(_.symbol != 'T').map(_.length).sum
val newClipplingPosition = clippingPosition - minus
```